contrib/drivers/hpe/nimble/nimble.go

```diff
@@ -0,0 +1,346 @@
+// Copyright 2019 The OpenSDS Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nimble
+
+import (
+	"fmt"
+	"math"
+
+	log "github.com/golang/glog"
+	. "github.com/opensds/opensds/contrib/drivers/utils/config"
+	"github.com/opensds/opensds/pkg/model"
+	pb "github.com/opensds/opensds/pkg/model/proto"
+	"github.com/opensds/opensds/pkg/utils/config"
+	"github.com/satori/go.uuid"
+)
+
+const (
+	DefaultConfPath = "/etc/opensds/driver/hpe_nimble.yaml"
+	NamePrefix      = "opensds"
+)
+
+type Config struct {
+	AuthOptions `yaml:"authOptions"`
+	Pool        map[string]PoolProperties `yaml:"pool,flow"`
+}
+
+type Driver struct {
+	conf   *Config
+	client *NimbleClient
+}
+
+func (d *Driver) Setup() (err error) {
+
+	conf := &Config{}
+	d.conf = conf
+	path := config.CONF.OsdsDock.Backends.HpeNimble.ConfigPath
+	if "" == path {
+		path = DefaultConfPath
+	}
+	Parse(conf, path)
+
+	d.client, err = NewClient(&d.conf.AuthOptions)
+	if err != nil {
+		log.Errorf("%v: get new client failed.\n%v", DriverName, err)
+		return err
+	}
+
+	return nil
+}
+
+func (d *Driver) Unset() error {
+	return nil
+}
+
+func (d *Driver) CreateVolume(opt *pb.CreateVolumeOpts) (*model.VolumeSpec, error) {
+	log.Infof("%v: try to create volume...", DriverName)
+
+	poolId, err := d.client.GetPoolIdByName(opt.GetPoolName())
+	if err != nil {
+		return nil, err
+	}
+	lun, err := d.client.CreateVolume(poolId, opt)
+	if err != nil {
+		log.Errorf("%v: create Volume Failed: %v", DriverName, err)
+		return nil, err
+	}
+	log.Infof("%v: create volume success.", DriverName)
+	return &model.VolumeSpec{
+		BaseModel: &model.BaseModel{
+			Id: opt.GetId(),
+		},
+		Name:             opt.GetName(),
+		Size:             lun.Size / int64(math.Pow(1024, 3)),
+		Description:      lun.Description,
+		AvailabilityZone: opt.GetAvailabilityZone(),
+		Metadata: map[string]string{
+			"Group":  lun.OwnedByGroup,
+			"Iqn":    lun.TargetName,
+			"LunId":  lun.Id,
+			"PoolId": poolId,
+		},
+	}, nil
+}
+
+func (d *Driver) DeleteVolume(opt *pb.DeleteVolumeOpts) error {
```

Comment:
Will doing explicit check if the volume is in use or not before actually trying to delete it make sense? Or the storage will throw an error if device is busy while deleting volume? And we are relying on that only

Reply:
When deleting volume, before that we have to offline that volume. I set "force option" in that section. So I removed this option because storage will be returning errors when device is busy.
Thank you for your advice.

Reply:
Force option makes volume offline even though "device is busy". Therefore I remove that option. fixed.

Reply:
Ok. great